[DO NOT MERGE] frisby/issue 1701 demo sandbox exe

nfrisby opened the pull request:
THIS PR WILL NEVER BE MERGED

I'm only opening it in order to facilitate conversation with @sterraf, who's trying out my WIP demo branch.

nfrisby committed:
LeiosDemo202510: draft Leios mini protocols

I'm intentionally grouping this code poorly and avoiding comments in an attempt
to emphasize that this code is currently being written to enable
hacked-together prototypes, not as a contribution towards code that might run
on mainnet. (Even though these definitions are much more compatible with that
development work than is other code being written for this prototype.)

dnadales committed:
Add the initial version of the demo script

dnadales committed:
Add node-0 to the path of different files

... such as topology, database, and socket.

nfrisby committed:
leiosdemo202510: add this demo exe

nfrisby committed:
leiosdemo202510: combine via join

nfrisby committed:
leiosdemo202510: suggestive check on MsgLeiosBlockTxsRequest

dnadales committed:
Run a second Cardano node as downstream peer

nfrisby committed:
leiosdemo202510: add msgLeiosBlockRequest and also stub limit on popCount

nfrisby committed:
leiosdemo202510: relocate SQL definitions near use

nfrisby committed:
leiosdemo202510: avoid reversing long lists

bladyjoker committed:
Scaffold leios-mvd NixOS test with leios-node

bladyjoker committed:
Git ignore NixOS test history files and direnv

nfrisby committed:
leiosdemo202510: fallback for requests of many txs

nfrisby committed:
leiosdemo202510: avoid reversing lists 2

nfrisby committed:
leiosdemo202510: use map CBOR and handle MsgLeiosBlock

nfrisby committed:
leiosdemo202510: handle MsgLeiosBlockTxs

bladyjoker committed:
Implements immdb-node and Immutable DB NixOS systemd service

bladyjoker committed:
Add `--address` flag to `immdb-server` executable

bladyjoker committed:
Add `address` options to `immdb-server` NixOS service

bladyjoker committed:
DNS tweaks to Leios MVD and formatting

bladyjoker committed:
Add immutable db fixture (50 blocks)

bladyjoker committed:
Configure the private testnet such that it doesn't update the system start time

bladyjoker committed:
Add README.md for Leios MVD and an actual test scenario condition

nfrisby committed:
leiosdemo202510: merge ebBodies and ebClosures, use only points in CLI, improve column names

nfrisby committed:
leiosdemo202510: minor refactor

nfrisby committed:
leiosdemo202510: add fetch-decision and hash-txs commands

nfrisby committed:
leiosdemo202510: only use "point" in the table name ebPoints

nfrisby committed:
leiosdemo202510: add some of the missing Haddock

dnadales committed:
Add a script to analyze the two node log files

Calculates the propagation latency for each block by measuring
the time difference between the completion of the block fetch on the
upstream node (node 0) and the completion of the block fetch on the
downstream node (node 1).

dnadales committed:
Hook the python script to the demo script

dnadales committed:
Add requirements.txt for log_parser.py

bladyjoker committed:
Nixify the Leios demo environment

bladyjoker committed:
Appliy black python formatter

bladyjoker committed:
Add Leios demo test data

bladyjoker committed:
Nixify the Leios demo environment (#1724)

bladyjoker committed:
Add the IPython notebook for interactive analysis

Run `jupyter-lab` in the provided Nix shell

dnadales committed:
Thread the onset of the reference slot through the immdb-server code

dnadales committed:
Pass a function that abstracts away slot delay calculation

bladyjoker committed:
Freeze the Python environment from Nix into requirements.txt

$ pip freeze > requirements.txt

nfrisby committed:
leiosdemo202510: redo the fetch iteration logic and MsgLeiosBlockTxs now also updates that same state

nfrisby committed:
leiosdemo202510: add the first bash script that mimics requests and replies

nfrisby committed:
leiosdemo202510: add tx sharing to the generator

dnadales committed:
Implement mkGetSlotDelay

nfrisby committed:
leiosdemo202510: fix off-by-one sterraf noticed

bladyjoker committed:
Remove unnecessary venvShellHook

bladyjoker committed:
Merge branch 'bladyjoker/leios-mvd-nixostests' into 1709-initial-demo-setup

bladyjoker committed:
Leios MVD setup with NixOS tests (#1718)

# Description

Leios Minimum Viable Demo branch with NixOS test based setup for spawning reproducible and/or interactive demo scenarios

bladyjoker committed:
Add new options for ImmDB NixOS service

nfrisby committed:
leiosdemo202510: some renamings, add MkLeiosBlockRequest ctor

nfrisby committed:
leiosdemo202510: some renamings, add EB body requests etc

nfrisby committed:
leiosdemo202510: small polishing pass

nfrisby commented on myManifest.json:
```diff
@@ -0,0 +1,55 @@
+[
```
@sterraf Here's an example `myManifest.json` file.

nfrisby committed:
leiosdemo202510: also insert into TxCache

nfrisby committed:
leiosdemo202510: do not fetch txs already in TxCache

nfrisby committed:
leiosdemo202510: add cache-copy command

nfrisby committed:
leiosdemo202510: rename binder to nickname in myManifest.json

bladyjoker committed:
Set all followert instructions to blocking

There's many rollback events that happen on the last block.

bladyjoker committed:
Cosmetic changes to aid my own comprehension

bladyjoker committed:
Switch to using plotly for charting events

nfrisby committed:
Merge remote-tracking branch 'origin/1709-initial-demo-setup' into nfrisby/issue-1701-demo-sandbox-exe

nfrisby committed:
leiosdemo202510: add mini protocol stubs to ImmDB server

bladyjoker committed:
Delay throwing ReachedImmutableTip to allow clients to fetch blocks

bladyjoker committed:
Adds journal logs from leios-node and adjust up the Python scripts

dnadales committed:
Add arguments for the anchor slot and its corresponding onset

dnadales committed:
Add support for POSIX timestamps

dnadales committed:
Calculate slot onset time for each slot in the logs

dnadales committed:
Pass anchor slot and its onset to the log_parser script

nfrisby committed:
leiosdemo202510: added generalized LeiosNotify peers to ouroboros-consensus-diffusion

dnadales committed:
Add a sanity check for the slot onset calculation

dnadales committed:
Make the script generate a scatter plot

nfrisby committed:
WIP

nfrisby committed:
WIP the script

nfrisby committed:
WIP my TODO list

nfrisby committed:
Revert "WIP the script"

This reverts commit 9571dbf669465d9ced67a8174c4845087d2fdee4.

nfrisby committed:
Merge remote-tracking branch 'origin/1709-initial-demo-setup' into nfrisby/issue-1701-demo-sandbox-exe

nfrisby committed:
WIP hacking on script file

nfrisby committed:
WIP fixup 291225921 LeiosSchedule parsing in immdb-server

nfrisby committed:
leiosdemo202510: use in-mem table instead of batching retrievals

nfrisby committed:
leiosdemo202510: polishing model exe

nfrisby committed:
leiosdemo202510: adding Leios state to NodeKernel

sterraf commented:
@nfrisby  After building and testing with `./script`, I seem to see that EBs are CBOR-encoded as indefinite maps. (I see that this saves at least  1 byte per reference.) 

Are you planning to modify the CDDL accordingly?

nfrisby committed:
leiosdemo202510: switch from RealPoint to bespoke LeiosPoint

nfrisby committed:
leiosdemo202510: add basic logic to LeiosNotify client

nfrisby committed:
leiosdemo202510: migrate EbId logic from the model exe

nfrisby committed:
fixup leiosdemo202510: LeiosNotify

nfrisby committed:
leiosdemo202510: add LeiosFetch to NodeToNode

nfrisby committed:
fixup leiosdemo202510: LeiosNotify

nfrisby committed:
leiosdemo202510: add generalized LeiosFetch pipelined client

nfrisby committed:
leiosdemo202510: switch requestsToSend to TVar, for convenience

nfrisby committed:
leiosdemo202510: use MutVar instead of MVar for drainThePipe

nfrisby committed:
leiosdemo202510: polishing demo script

nfrisby committed:
leiosdemo202510: add LeiosFetch to immdb-server

nfrisby committed:
leiosdemo202510: fixup LeiosNotify client

nfrisby committed:
leiosdemo202510: add LeiosFetch server to node

nfrisby committed:
leiosdemo202510: more direct LeiosNotify server

nfrisby committed:
leiosdemo202510: add fetch decision logic to node

nfrisby committed:
leiosdemo202510: now sending LeiosFetch requests, but ignoring responses

nfrisby committed:
leiosdemo202510: update model exe EB codec to match

nfrisby committed:
leiosdemo202510: now processing MsgLeiosBlock, but ignoring MsgLeiosBlockTxs

nfrisby committed:
WIP one minute copies

nfrisby committed:
WIP bugfix, see longer message

This commit fixes two bugs:

1) The JOIN-based UPDATE for copying from txCache to ebTxs was taking ~72
   seconds. I replaced it with just a single-row UPDATE per index, and that
   seems better. I'm surprised, but maybe the bulk update query SQL can be
   tuned another time.

2) `packRequests` had a bug where it'd fail to request one tx whenever it had
   to flush due to size. Instead of being the first tx in the next request, it
   was simply never requested.

It also leaves all the debug stuff in, because I need to go to bed.

nfrisby committed:
WIP disable the most noisy traces

nfrisby committed:
WIP fixup warnings

nfrisby committed:
leiosdemo202510: drop the netstat stuff

This originated when I was debugging busy ports. Turns out I had some docker
containers running that I wasn't aware of (and they kept restarting). Once I
eventually noticed and dropped those containers, there were no more
unexpectedly-busy ports.

nfrisby committed:
leiosdemo202510: consolidate magic numbers (absorbing maxIngressQueue)

nfrisby committed:
leiosdemo202510: consolidate ebId magic number

nfrisby committed:
leiosdemo202510: also generate typical schedule.json file

nfrisby committed:
leiosdemo202510: track outstanding EbTxs count per EB

nfrisby committed:
leiosdemo202510: more interpretable packing of ebIds, for now

nfrisby committed:
leiosdemo202510: LeiosNotify server in node, but DB locked :/

nfrisby committed:
WIP HasCallStack for db

nfrisby committed:
WIP node-0 freezing???

nfrisby committed:
WIP change host-addr?

nfrisby committed:
WIP fix missing finalize and dieStack = dieStack :'(

nfrisby committed:
leiosdemo202510: LeiosFetch for downstream too

nfrisby committed:
leiosdemo202510: add a 2nd node-under-test, but not seeing mixed traffic?

nfrisby committed:
Revert "leiosdemo202510: add a 2nd node-under-test, but not seeing mixed traffic?"

This reverts commit ea74e1b398a3739d00fda4a8869c078372348381.

nfrisby committed:
leiosdemo202510: add missing VACUUM

nfrisby committed:
leiosdemo202510: do not elide CompletedBlockFetch

nfrisby committed:
leiosdemo202510: hack leios-october-script.sh and log_parser.py to show text latencies

nfrisby committed:
leiosdemo202510: disable scatter plot for now

nfrisby committed:
leiosdemo202510: polishing the leios-october-script.sh

nfrisby committed:
leiosdemo202510: add README-Leios-October-demo.md and demoManifest.json

nfrisby committed:
leiosdemo202510: add some documentation to README-Leios-October-demo.md

nfrisby committed:
leiosdemo202510: fix typos in README-Leios-October-demo.md

sterraf commented:
> FYI @sterraf I just pushed up this file https://github.com/IntersectMBO/ouroboros-consensus/blob/nfrisby/issue-1701-demo-sandbox-exe/README-Leios-October-demo.md , which might be useful to read depending on what you're trying to do with this branch.

I got stuck with an already very stale commit, so I just took some data generated by the demo at that point and used it in my own tests. Not sure if I'll be able to keep doing experiments after tomorrow's call 😕

In any case, thanks a lot!

nfrisby committed:
leiosdemo202510: update now that the simplest first experiment is working

nfrisby commented:
Closed as stale. Superseded by `leios-prototype` branch.
